fv3core/testing/validation.py

```diff
@@ -0,0 +1,87 @@
+):
+    """
+    Convert a model class into one that sets nans on non-validated outputs,
+    and gives a helper function to retrieve the output subset we want to validate.
+
+    Using this ensures that if these non-validated values are ever used, a test
+    will fail because an output will have NaN.
+    """
+
+    class SelectivelyValidated:
+        """
+        Wrapper class that sets non-validated outputs to nan, and gives a helper
+        function to retrieve the output subset to be validated.
+        """
+
+        def __init__(self, *args, **kwargs):
+            self.wrapped = cls(*args, **kwargs)
+            origin, domain = origin_domain_func(self.wrapped)
+            self._validation_slice = tuple(
+                slice(start, start + n) for start, n in zip(origin, domain)
+            )
+            self.origin = origin
+            self.domain = domain
+            self._all_argument_names = tuple(
+                inspect.getfullargspec(self.wrapped).args[1:]
+            )
+            assert "self" not in self._all_argument_names
+            self._selective_argument_names = selective_arg_names
+
+        def __call__(self, *args, **kwargs):
+            kwargs.update(self._args_to_kwargs(args))
+            self.wrapped(**kwargs)
+            self._set_nans(kwargs)
+
+        def _args_to_kwargs(self, args):
+            return dict(zip(self._all_argument_names, args))
+
+        def subset_output(self, varname: str, output: np.ndarray) -> np.ndarray:
+            """
+            Given an output array, return the slice of the array which we'd
+            like to validate against reference data
+            """
+            if varname in self._selective_argument_names:
+                output = output[self._validation_slice]
+            return output
+
+        def _set_nans(self, kwargs):
+            for name in set(kwargs.keys()).intersection(self._selective_argument_names):
+                array = kwargs[name]
+                validation_data = np.copy(array[self._validation_slice])
+                array[:] = np.nan
+                array[self._validation_slice] = validation_data
+
+    return SelectivelyValidated
+
+
+def get_update_height_on_d_grid_selective_domain(
+    instance,
+) -> Tuple[Tuple[int, ...], Tuple[int, ...]]:
+    origin = instance.grid.compute_origin()
+    domain = instance.grid.domain_shape_compute(add=(0, 0, 1))
+    return origin, domain
+
+
+def enable_selective_validation():
+    # to enable selective validation for a new class, add a new monkeypatch
+    # this should require only a new function for (origin, domain)
+    # note we have not implemented disabling selective validation once enabled
+    fv3core.stencils.updatedzd.UpdateHeightOnDGrid = get_selective_class(
+        fv3core.stencils.updatedzd.UpdateHeightOnDGrid,
+        ["height", "zh"],  # must include both function and savepoint names
+        get_update_height_on_d_grid_selective_domain,
+    )
```

Comment:
there is a tradeoff here, I think it is a good one (duplicating class information but getting validation code out of the main model code)